content/advent-2017/messaging-framework.md

```diff
@@ -0,0 +1,709 @@
+```go
+func (s *TcpServer) SendDataByClientId(clientUid string, data []byte) error{
+	if s.connections[clientUid]!=nil {
+		return s.connections[clientUid].Send(data)
+	} else {
+		return errors.New(fmt.Sprint("no connection with uid ", clientUid))
+	}
+
+	return nil
+}
+
+func (s *TcpServer) SendDataByDeviceUid(deviceUid string, data []byte) error{
+	for k := range s.connections {
+		if s.connections[k].DeviceUid == deviceUid {
+			return s.connections[k].Send(data)
+		}
+	}
+	return errors.New(fmt.Sprint("no connection with deviceUid ", deviceUid))
+}
+```
+
+## API
+
+We need to create structs for the API that the tcp clients use, and the API for the messages sent to/from the messages broker.  
+For the TCP clients:
+```go
+type DeviceRequest struct {
+	Action string `json:"action"`
+	DeviceUid string `json:"deviceUid"`
+	Uid string `json:"uid"`
+	Data map[string]interface{} `json:"data"`
+}
+type DeviceResponse struct {
+	Action string `json:"action"`
+	Uid string `json:"uid"`
+	Data map[string]interface{} `json:"data"`
+	Status string `json:"status"`
+	ErrorCode string `json:"errorCode"`
+	ErrorMessage string `json:"errorMessage"`
+}
+```
+For the message broker:
+```go
+type ServerRequest struct{
+	DeviceRequest DeviceRequest `json:"deviceRequest"`
+	ServerId string `json:"serverId"` // unique identifier for each server in case of having more than 1 server
+	ClientId string `json:"clientId"`
+}
+type ServerResponse struct {
+	DeviceResponse DeviceResponse `json:"deviceResponse"`
+	ServerId string `json:"serverId"` // unique identifier for each server in case of having more than 1 server
+	ClientId string `json:"clientId"`
+	DeviceUid string `json:"deviceUid"`
+}
+```
+## Configurations
+I use ```.yml``` files for configurations that change between environments. Here is how to model them and parse from the ```.yml``` file.
+```go
+func InitConfig(configPath string) {
+	if conf == nil {
+		filename, _ := filepath.Abs(configPath)
+		log.Println("trying to read file ", filename)
+		yamlFile, err := ioutil.ReadFile(filename)
+		var confi Configuration
+		err = yaml.Unmarshal(yamlFile, &confi)
+		if err != nil {
+			log.Println(err)
+			panic(err)
+		}
+		conf = &confi
+	}
+}
+
+func Get() *Configuration {
+	return conf
+}
+
+type Configuration struct {
+	BrokersList []string  `yaml:"brokers_list"`
+	ProducerTopic		string `yaml:"producer_topic"`
+	ConsumerTopics		[]string `yaml:"consumer_topics"`
+	ConsumerGroupId     string `yaml:"consumer_group_id"`
+}
+```
+And here is an example for a ```yml``` config file:
+```yaml
+brokers_list:
+  - "localhost:9092"
+producer_topic: "tcp_layer_messages"
+consumer_topics:
+  - "workers_layer_messages"
+consumer_group_id: "id-1"
+```
+## Main function - putting it all together
+Here obtain and manages all the other components in this system. It will include the TCP server that holds an array of TCP clients, and a connection to the Kafka broker for consuming and sending messages to it.
```

Comment:
Here **we** obtain
s/manages/manage

Reply:
changed to : Obtains and manages all the other components in this system 👍